community/kernel/src/main/java/org/neo4j/kernel/impl/logging/StoreLogService.java

```diff
@@ -149,40 +167,81 @@ private StoreLogService( LogProvider userLogProvider,
         }
         else
         {
-            RotatingFileOutputStreamSupplier rotatingSupplier = new RotatingFileOutputStreamSupplier( fileSystem, internalLog,
-                    internalLogRotationThreshold, internalLogRotationDelay, maxInternalLogArchives,
-                    rotationExecutor, new RotatingFileOutputStreamSupplier.RotationListener()
+            RotatingFileOutputStreamSupplier rotatingSupplier = new RotatingFileOutputStreamSupplier(
+                    fileSystem,
+                    internalLog,
+                    builder.internalLogRotationThreshold,
+                    builder.internalLogRotationDelay,
+                    builder.maxInternalLogArchives,
+                    builder.rotationExecutor,
+                    new RotatingFileOutputStreamSupplier.RotationListener()
             {
                 @Override
                 public void outputFileCreated( OutputStream newStream, OutputStream oldStream )
                 {
                     FormattedLogProvider logProvider = internalLogBuilder.toOutputStream( newStream );
                     logProvider.getLog( StoreLogService.class ).info( "Opened new internal log file" );
-                    rotationListener.accept( logProvider );
+                    builder.rotationListener.accept( logProvider );
                     logProvider.getLog( StoreLogService.class ).info( "Rotated internal log file" );
                 }
             } );
             internalLogProvider = internalLogBuilder.toOutputStream( rotatingSupplier );
-            this.closeable = rotatingSupplier;
+            closeInShutdown = rotatingSupplier;
         }
+
+        // Configure the log providers to be asynchronous, if user wanted that behavior.
+        LogProvider userLogProvider = builder.userLogProvider;
+        Runnable runInInit = Runnables.NO_OP;
+        if ( builder.asyncExecutor != null )
+        {
+            // Wraps each log provider in an AsyncLogProvider and schedules the flushing of log events in an executor,
+            // where the scheduling into the executor takes place during init().
+            final AsyncLogProvider asyncUserLogProvider = new AsyncLogProvider( userLogProvider );
+            final AsyncLogProvider asyncInternalLogProvider = new AsyncLogProvider( internalLogProvider );
+
+            // Reassign these log providers so that the async versions are used below.
+            userLogProvider = asyncUserLogProvider;
+            internalLogProvider = asyncInternalLogProvider;
+
+            // Close in correct order, i.e: log providers and THEN the original closeable
+            // (the underlying log output stream).
+            runInInit = new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    builder.asyncExecutor.execute( asyncUserLogProvider );
+                    builder.asyncExecutor.execute( asyncInternalLogProvider );
```

Comment:
This looks like we will get two different log processing threads, one for each of the two logs we have.
Since we (most likely) only have one I/O subsystem, would it not be better to process both event streams on the same thread?

That would mean enqueueing the events for both logs into the same `AsyncEvents` queue (since the events know which log they are writing to already, this should be fine).